atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleDdlTable.java

```diff
@@ -146,10 +149,40 @@ private boolean overflowTableHasMigrated() {
         return config.overflowMigrationState() == OverflowMigrationState.FINISHED;
     }
 
-    private boolean needsOverflow(byte[] tableMetadata) {
-        TableMetadata metadata = TableMetadata.BYTES_HYDRATOR.hydrateFromBytes(tableMetadata);
-        return (metadata != null)
-                && (metadata.getColumns().getMaxValueSize() > AtlasDbConstants.ORACLE_OVERFLOW_THRESHOLD);
+    private boolean needsOverflow(TableMetadata metadata) {
+        return metadata.getColumns().getMaxValueSize() > AtlasDbConstants.ORACLE_OVERFLOW_THRESHOLD;
+    }
+
+    /*
+     * Determine the optimal Index Organized Table compression setting.
+     * When enabling compression on an IOT, we can instruct Oracle to compress at most one
+     * fewer columns than the number of columns that make up the PK for the IOT (i.e. the IOT we create have a PK of
+     * {row, col, ts} so we can compress 0, 1 or 2 columns - nothing, row, or row & col. It wouldn't make sense to
+     * compress all three columns, because necessarily there can be multiple rows that share all three values.
+     *
+     * When compression is enabled in an IOT, the block header adds a table the captures the values of the
+     * compressed columns. The actual rows in the blocks then refer back to entries in the header table to capture
+     * their values for the compressed columns. This extra header table and the slot references for data rows consume
+     * space in the block, so misuse of compression results in extra storage being consumed.
+     *
+     * Compression only makes sense when we expect that the { row } or { row, col } values will be reused:
+     * - If the table's sweep strategy is CONSERVATIVE, then every { row, col } will have at least two rows,
+     *   the current value (which may have null val if the value is deleted) and a sentinel row with ts = -1. Clearly
+     *   compression of both { row, col } is valuable.
+     * - If the sweep strategy is not CONSERVATIVE,  there will be reuse of the { row } if there is more than one named
+     *   column, or the table uses dynamic columns, which would be generally be expected to have multiple row values. In
+     *   this case, compression 1 is appropriate so { row } only needs to be store once.
+     * - Finally, if the sweep strategy is not CONSERVATIVE, there are no sentinels, and if there is only one named
+     *   column then there will be no rows that share { row } so compression should be disabled.
+     */
+    private int getOptimalIndexCompression(TableMetadata metadata) {
+        if (metadata.getSweepStrategy() == SweepStrategy.CONSERVATIVE) {
+            return 2;
+        } else if (!metadata.getColumns().hasDynamicColumns() || metadata.getColumns().getNamedColumns().size() > 1) {
+            return 1;
+        } else {
+            return 0;
+        }
```

Comment:
There's a bunch of tests failing, likely due to `getNamedColumns()` returning null (we should also apply `@Nullable` to `ColumnMetadataDescription` appropriately so that `NullAway` could catch this at compile time when AtlasDB enables it). I also don't think the implementation matches the comments above for dynamic columns cases.



```
java.lang.NullPointerException
	at com.palantir.atlasdb.keyvalue.dbkvs.impl.oracle.OracleDdlTable.getOptimalIndexCompression(OracleDdlTable.java:181)
	at com.palantir.atlasdb.keyvalue.dbkvs.impl.oracle.OracleDdlTable.create(OracleDdlTable.java:96)
	at com.palantir.atlasdb.keyvalue.dbkvs.impl.DbKvs.lambda$createTable$34(DbKvs.java:1174)
	at com.palantir.atlasdb.keyvalue.dbkvs.impl.DbKvs.runDdl(DbKvs.java:1330)
	at com.palantir.atlasdb.keyvalue.dbkvs.impl.DbKvs.createTable(DbKvs.java:1173)
	at com.palantir.atlasdb.keyvalue.impl.AbstractKeyValueService.createTables(AbstractKeyValueService.java:119)
	at com.palantir.atlasdb.keyvalue.api.AutoDelegate_KeyValueService.createTables(AutoDelegate_KeyValueService.java:144)
	at com.palantir.atlasdb.transaction.impl.TransactionTestSetup.setUp(TransactionTestSetup.java:144)
```

```suggestion
        if (metadata.getSweepStrategy() == SweepStrategy.CONSERVATIVE) {
            // the table's sweep strategy is CONSERVATIVE, then every { row, col } will have at least two rows
            return 2;
        }
        ColumnMetadataDescription columns = metadata.getColumns();
        if (columns != null) {
            // there will be reuse of the { row } if there is more than one named column
            // or the table uses dynamic columns, which would be generally be expected to have multiple row values
            if (columns.hasDynamicColumns()) {
                return 1;
            }
            Set<NamedColumnDescription> namedColumns = columns.getNamedColumns();
            if (namedColumns != null && namedColumns.size() > 1) {
                return 1;
            }
        }

        return 0;
```